File: shapes.py
# -*- coding: utf-8 -*-

from abc import ABC, abstractmethod

class Shape(ABC):
	def __init__(self, sides):
		self.sides = sides
	
	def get_sides(self):
		return sides

	@abstractmethod
	def display(self):
		pass

class Circle(Shape):
	def __init__(self, radius, center):
		Shape.__init__(self, float('Inf'))
		self.radius = radius
		self.center = center
	
	def display(self):
		print("●")

class Triangle(Shape):
	def __init__(self, center):
		Shape.__init__(self, 3)
		self.center = center

	def display(self):
		print('▲')

class Square(Shape):
	def __init__(self, center):
		Shape.__init__(self, 4)
		self.center = center
	
	def display(self):
		print('◼')


database = [
	Square((9,1)),
	Triangle((9,1)),
	Circle(0.1, (1,1)),
	Square((1,1)),
	Square((1,2)),
	Circle(0.2, (2,1)),
	Circle(0.3, (1,5)),
	Circle(0.4, (1,1)),
	Square((1,3)),
	Triangle((1,3)),
	Square((9,1)),
	Square((9,1)),
	Circle(0.5, (2,1)),
	Triangle((1,1)),
	Triangle((1,2)),
	Circle(0.6, (1,20)),
	Circle(0.7, (8,1)),
	Square((1,1)),
	Square((1,2)),
	Triangle((1,1)),
	Triangle((1,2)),
	Triangle((1,3)),
	Square((1,3)),
] 

sorted_shapes = sorted(database, key=lambda x: x.sides, reverse=True)

for shape in sorted_shapes:
	shape.display()

File: shapes.rb
class Shape
  attr_reader :sides, :center

  def initialize(sides, center)
    @sides = sides
    @center = center
  end

  def display
    raise "should be implemented in subclass"
  end
end

class Circle < Shape
  attr_reader :radius
  def initialize(radius, center)
    super(Float::INFINITY, center)
    @radius = radius
  end

  def display
    puts "●"
  end
end

class Triangle < Shape
  def initialize(center)
    super(3, center)
  end

  def display
    puts "▲"
  end
end

class Square < Shape
  def initialize(center)
    super(4, center)
  end
  
  def display
    puts "◼"
  end
end

database = [
  Square.new([9,1]),
	Triangle.new([9,1]),
	Circle.new(0.1, [1,1]),
	Square.new([1,1]),
	Square.new([1,2]),
	Circle.new(0.2, [2,1]),
	Circle.new(0.3, [1,5]),
	Circle.new(0.4, [1,1]),
	Square.new([1,3]),
	Triangle.new([1,3]),
	Square.new([9,1]),
	Square.new([9,1]),
	Circle.new(0.5, [2,1]),
	Triangle.new([1,1]),
	Triangle.new([1,2]),
	Circle.new(0.6, [0,20]),
	Circle.new(0.7, [8,1]),
	Square.new([1,1]),
	Square.new([1,2]),
	Triangle.new([1,1]),
	Triangle.new([1,2]),
	Triangle.new([1,3]),
	Square.new([1,3]),
]


sorted = database.sort_by {|obj| obj.sides}

for s in sorted do
  s.display
end
